File: ios/Plugin/Utils/UtilsSecret.swift
//
//  UtilsSecret.swift
//  Plugin
//
//  Created by  Quéau Jean Pierre on 04/05/2021.
//  Copyright © 2021 Max Lynch. All rights reserved.
//

import Foundation

enum UtilsSecretError: Error {
    case prefixPassphrase(message: String)
    case setPassphrase(message: String)
    case changePassphrase(message: String)
    case setEncryptionSecret(message: String)
    case changeEncryptionSecret(message: String)
    case clearEncryptionSecret(message: String)
    case checkEncryptionSecret(message: String)
}

let oldAccount: String = "CapacitorSQLitePlugin"

class UtilsSecret {

    // MARK: - IsPassphrase

    class func isPassphrase(account: String) throws -> Bool {

        if !getPassphrase(account: account).isEmpty {
            return true
        }
        if !getPassphrase(account: oldAccount).isEmpty {
            let passphrase = getPassphrase(account: oldAccount)
            do {
                try setPassphrase(account: account, passphrase: passphrase)
                return true
            } catch UtilsSecretError.prefixPassphrase(let message) {
                throw UtilsSecretError.changePassphrase(message: message)
            }
        }
        return false
    }

    // MARK: - GetPassphrase

    class func getPassphrase(account: String) -> String {
        let kcw = KeychainWrapper()
        if let password = try? kcw.getGenericPasswordFor(
            account: account,
            service: "unlockSecret") {
            return password
        }
        return ""
    }

    // MARK: - SetPassphrase

    class func setPassphrase(account: String, passphrase: String) throws {
        let kcw = KeychainWrapper()
        do {
            try kcw.storeGenericPasswordFor(
                account: account,
                service: "unlockSecret",
                password: passphrase)
        } catch let error as KeychainWrapperError {
            var msg: String = "SetPassphrase command failed :"
            msg.append(" \(error.localizedDescription)")
            throw UtilsSecretError.setPassphrase(message: msg)

        } catch {
            let msg: String = "SetPassphrase command failed :"
            throw UtilsSecretError.setPassphrase(message: msg)
        }

    }

    // MARK: - ValidatePassphrase

    class func validatePassphrase(account: String, passphrase: String) -> Bool {
        var ret: Bool = false
        let currentPassphrase = getPassphrase(account: account)
        if passphrase == currentPassphrase {
            ret = true
        }
        return ret
    }

    // MARK: - ChangePassphrase

    class func changePassphrase(account: String, oldPassphrase: String,
                                passphrase: String) throws -> Bool {
        guard validatePassphrase(account: account,
                                 passphrase: oldPassphrase) == true else { return false }
        do {
            try setPassphrase(account: account, passphrase: passphrase)
            return true
        } catch UtilsSecretError.setPassphrase(let message) {
            throw UtilsSecretError.changePassphrase(message: message)
        }
    }

    // MARK: - SetEncryptionSecret

    // swiftlint:disable function_body_length
    // swiftlint:disable cyclomatic_complexity
    class func setEncryptionSecret(prefix: String, passphrase: String,
                                   databaseLocation: String) throws {
        do {
            if prefix.isEmpty {
                let msg: String = "keychain prefix must not be empty"
                throw UtilsSecretError.setEncryptionSecret(message: msg)
            }
            if passphrase.isEmpty {
                let msg: String = "passphrase must not be empty"
                throw UtilsSecretError.setEncryptionSecret(message: msg)
            }
            // store encrypted passphrase
            let account = "\(prefix)_\(oldAccount)"
            if !getPassphrase(account: account).isEmpty {
                let msg: String = "passphrase already stored in keychain"
                throw UtilsSecretError.setEncryptionSecret(message: msg)
            }
            try setPassphrase(account: account, passphrase: passphrase)

            // get the list of databases
            let databaseURL: URL = try UtilsFile.getDatabasesUrl().absoluteURL
            var isDir = ObjCBool(true)
            if FileManager.default.fileExists(atPath: databaseURL.relativePath,
                                              isDirectory: &isDir) && isDir.boolValue {
                let dbList: [String] = try UtilsFile
                    .getFileList(path: databaseURL.relativePath, ext: ".db")
                for file: String in dbList {
                    let state: State = UtilsSQLCipher
                        .getDatabaseState(databaseLocation: databaseLocation,
                                          databaseName: file, account: account)
                    if state.rawValue == "ENCRYPTEDGLOBALSECRET" {
                        let globalData: GlobalSQLite = GlobalSQLite()
                        let password: String = globalData.secret

                        let dbPath: String  = try UtilsFile
                            .getFilePath(databaseLocation: databaseLocation,
                                         fileName: file)
                        try UtilsSQLCipher.changePassword(filename: dbPath,
                                                          password: password,
                                                          newpassword: passphrase)
                    } else if state.rawValue == "ERROR" {
                        let msg: String = "Closing \(file) failed"
                        throw UtilsSecretError.setEncryptionSecret(message: msg)
                    } else if state.rawValue == "DOESNOTEXIST"
                                || state.rawValue == "UNKNOWN" {
                        let msg: String = "State for: \(file) not correct"
                        throw UtilsSecretError.setEncryptionSecret(message: msg)
                    }
                }
            } else {
                let msg: String = "No Database folder"
                throw UtilsSecretError.setEncryptionSecret(message: msg)
            }
        } catch UtilsSecretError.setPassphrase(let message) {
            throw UtilsSecretError.setEncryptionSecret(message: message)
        } catch UtilsFileError.getFilePathFailed {
            let msg: String = "Failed when getting a file path"
            throw UtilsSecretError.setEncryptionSecret(message: msg)
        } catch UtilsSQLCipherError.changePassword(let message) {
            throw UtilsSecretError.setEncryptionSecret(message: message)
        }

    }
    // swiftlint:enable cyclomatic_complexity
    // swiftlint:enable function_body_length

    // MARK: - ChangeEncryptionSecret

    // swiftlint:disable function_body_length
    class func changeEncryptionSecret(prefix: String, passphrase: String,
                                      oldPassphrase: String,
                                      databaseLocation: String) throws {
        do {
            if prefix.isEmpty {
                let msg: String = "Keychain prefix must not " +
                    "be empty"
                throw UtilsSecretError.changeEncryptionSecret(message: msg)
            }
            if passphrase.isEmpty ||  oldPassphrase.isEmpty {
                let msg: String = "Passphrase and/or oldpassphrase must not " +
                    "be empty"
                throw UtilsSecretError.changeEncryptionSecret(message: msg)
            }
            let account = "\(prefix)_\(oldAccount)"
            guard try isPassphrase(account: account) == true else {
                let msg: String = "Encryption secret has not been set"
                throw UtilsSecretError.changeEncryptionSecret(message: msg)
            }
            guard validatePassphrase(account: account,
                                     passphrase: oldPassphrase) == true else {
                let msg: String = "Given oldpassphrase is wrong"
                throw UtilsSecretError.changeEncryptionSecret(message: msg)
            }
            // get the list of databases from the database folder
            let databaseURL: URL = try UtilsFile
                .getFolderURL(folderPath: databaseLocation).absoluteURL
            var isDir = ObjCBool(true)
            if FileManager.default.fileExists(atPath: databaseURL.relativePath,
                                              isDirectory: &isDir) && isDir.boolValue {
                let dbList: [String] = try UtilsFile
                    .getFileList(path: databaseURL.relativePath, ext: ".db")
                for file: String in dbList {
                    let state: State = UtilsSQLCipher
                        .getDatabaseState(databaseLocation: databaseLocation,
                                          databaseName: file, account: account)
                    if state.rawValue == "ENCRYPTEDSECRET" {
                        let dbPath: String  = try UtilsFile
                            .getFilePath(databaseLocation: databaseLocation,
                                         fileName: file)
                        try UtilsSQLCipher.changePassword(filename: dbPath,
                                                          password: oldPassphrase,
                                                          newpassword: passphrase)
                    } else if state.rawValue == "ERROR" {
                        let msg: String = "Closing \(file) failed"
                        throw UtilsSecretError.setEncryptionSecret(message: msg)
                    } else if state.rawValue == "DOESNOTEXIST"
                                || state.rawValue == "UNKNOWN"
                                || state.rawValue == "ENCRYPTEDGLOBALSECRET" {
                        let msg: String = "State for: \(file) not correct"
                        throw UtilsSecretError.setEncryptionSecret(message: msg)
                    }

                }

            } else {
                let msg: String = "No Database folder"
                throw UtilsSecretError.setEncryptionSecret(message: msg)
            }

            // store encrypted passphrase
            try setPassphrase(account: account, passphrase: passphrase)

        } catch UtilsSecretError.setPassphrase(let message) {
            throw UtilsSecretError.setEncryptionSecret(message: message)
        }

    }
    // swiftlint:enable function_body_length

    // MARK: - ClearEncryptionSecret

    class func clearEncryptionSecret(prefix: String, databaseLocation: String) throws {
        do {
            if prefix.isEmpty {
                let msg: String = "keychain prefix must not be empty"
                throw UtilsSecretError.setEncryptionSecret(message: msg)
            }
            // clear encrypted passphrase
            let account = "\(prefix)_\(oldAccount)"
            if !getPassphrase(account: account).isEmpty {
                try setPassphrase(account: account, passphrase: "")
            }
        } catch UtilsSecretError.setPassphrase(let message) {
            throw UtilsSecretError.clearEncryptionSecret(message: message)
        }

    }

    // MARK: - CheckEncryptionSecret

    class func checkEncryptionSecret(prefix: String, passphrase: String) throws -> NSNumber {
        var ret: NSNumber = 0
        if prefix.isEmpty {
            let msg: String = "keychain prefix must not be empty"
            throw UtilsSecretError.checkEncryptionSecret(message: msg)
        }
        if passphrase.isEmpty {
            let msg: String = "passphrase must not be empty"
            throw UtilsSecretError.checkEncryptionSecret(message: msg)
        }
        // get encrypted passphrase
        let account = "\(prefix)_\(oldAccount)"
        let storedPassPhrase = getPassphrase(account: account)
        if storedPassPhrase.isEmpty {
            let msg: String = "no passphrase stored in keychain"
            throw UtilsSecretError.checkEncryptionSecret(message: msg)
        }
        if storedPassPhrase == passphrase {
            ret = 1
        }
        return ret
    }
}

File: android/src/main/java/com/getcapacitor/community/database/sqlite/SQLite/UtilsSecret.java
package com.getcapacitor.community.database.sqlite.SQLite;

import static com.getcapacitor.community.database.sqlite.SQLite.UtilsSQLCipher.State.DOES_NOT_EXIST;
import static com.getcapacitor.community.database.sqlite.SQLite.UtilsSQLCipher.State.ENCRYPTED_GLOBAL_SECRET;
import static com.getcapacitor.community.database.sqlite.SQLite.UtilsSQLCipher.State.ENCRYPTED_SECRET;
import static com.getcapacitor.community.database.sqlite.SQLite.UtilsSQLCipher.State.UNKNOWN;

import android.content.Context;
import android.content.SharedPreferences;
import android.text.TextUtils;
import java.io.File;
import net.sqlcipher.database.SQLiteDatabase;

public class UtilsSecret {

    private static final String TAG = UtilsFile.class.getName();
    private UtilsFile uFile = new UtilsFile();
    private GlobalSQLite globVar = new GlobalSQLite();
    private UtilsSQLCipher uCipher = new UtilsSQLCipher();

    private SharedPreferences sharedPreferences;
    private Context context;

    public UtilsSecret(Context context, SharedPreferences sharedPreferences) {
        this.context = context;
        this.sharedPreferences = sharedPreferences;
    }

    /**
     * SetEncryptionSecret
     * @param passphrase
     * @throws Exception
     */
    public void setEncryptionSecret(String passphrase) throws Exception {
        try {
            if (TextUtils.isEmpty(passphrase)) {
                String msg = "passphrase must not be empty";
                throw new Exception(msg);
            }
            // test if Encryption secret is already set
            String savedPassPhrase = getPassphrase();
            if (savedPassPhrase != null && savedPassPhrase.length() > 0) {
                throw new Exception("a passphrase has already been set ");
            }
            // Store encrypted passphrase in sharedPreferences
            setPassphrase(passphrase);

            // Get the list of databases
            String[] dbList = uFile.getListOfFiles(context);
            if (dbList.length > 0) {
                for (String dbName : dbList) {
                    File file = context.getDatabasePath(dbName);

                    UtilsSQLCipher.State state = uCipher.getDatabaseState(context, file, sharedPreferences, globVar);
                    // change password if encrypted with globVar.secret
                    if (state == ENCRYPTED_GLOBAL_SECRET) {
                        uCipher.changePassword(context, file, globVar.secret, passphrase);
                    } else if (state == DOES_NOT_EXIST || state == UNKNOWN) {
                        String msg = "State for: " + dbName + " not correct";
                        throw new Exception(msg);
                    }
                }
            }
        } catch (Exception e) {
            throw new Exception(e.getMessage());
        }
    }

    /**
     * ChangeEncryptionSecret
     * @param passphrase
     * @param oldPassphrase
     * @throws Exception
     */
    public void changeEncryptionSecret(String passphrase, String oldPassphrase) throws Exception {
        try {
            if (TextUtils.isEmpty(passphrase) || TextUtils.isEmpty(oldPassphrase)) {
                String msg = "Passphrase and/or oldpassphrase must not be empty";
                throw new Exception(msg);
            }
            // check the oldPassphrase
            String secret = getPassphrase();
            if (secret == null || secret.length() == 0) {
                String msg = "Encryption secret has not been set";
                throw new Exception(msg);
            } else if (!secret.equals(oldPassphrase)) {
                String msg = "Oldpassphrase is wrong secret";
                throw new Exception(msg);
            } else {
                // Get the list of databases
                String[] dbList = uFile.getListOfFiles(context);
                if (dbList.length > 0) {
                    for (String dbName : dbList) {
                        File file = context.getDatabasePath(dbName);

                        UtilsSQLCipher.State state = uCipher.getDatabaseState(context, file, sharedPreferences, globVar);
                        // change password if encrypted with oldPassphrase
                        if (state == ENCRYPTED_SECRET) {
                            uCipher.changePassword(context, file, oldPassphrase, passphrase);
                        } else if (state == DOES_NOT_EXIST || state == ENCRYPTED_GLOBAL_SECRET || state == UNKNOWN) {
                            String msg = "State for: " + dbName + " not correct";
                            throw new Exception(msg);
                        }
                    }
                }
                // Store the new encrypted passphrase in sharedPreferences
                setPassphrase(passphrase);
            }
        } catch (Exception e) {
            throw new Exception(e.getMessage());
        }
    }

    /**
     * ClearEncryptionSecret
     * @throws Exception
     */
    public void clearEncryptionSecret() throws Exception {
        try {
            // test if Encryption secret is already set
            String savedPassPhrase = getPassphrase();
            if (savedPassPhrase != null && savedPassPhrase.length() > 0) {
                // Clear encrypted passphrase in sharedPreferences
                clearPassphrase();
            }
        } catch (Exception e) {
            throw new Exception(e.getMessage());
        }
    }

    /**
     * CheckEncryptionSecret
     * @param passphrase
     * @throws Exception
     */
    public Boolean checkEncryptionSecret(String passphrase) throws Exception {
        Boolean ret = false;
        try {
            if (TextUtils.isEmpty(passphrase)) {
                String msg = "passphrase must not be empty";
                throw new Exception(msg);
            }
            // test if Encryption secret is already set
            String savedPassPhrase = getPassphrase();
            if (savedPassPhrase.isEmpty()) {
                throw new Exception("no passphrase stored  in sharedPreferences");
            }

            if (savedPassPhrase.equals(passphrase)) {
                ret = true;
            }
            return ret;
        } catch (Exception e) {
            throw new Exception(e.getMessage());
        }
    }

    public void setPassphrase(String passphrase) {
        sharedPreferences.edit().putString("secret", passphrase).apply();
    }

    public String getPassphrase() {
        String passphrase = sharedPreferences.getString("secret", "");
        return passphrase;
    }

    public void clearPassphrase() {
        sharedPreferences.edit().remove("secret").commit();
    }

    public Boolean isPassphrase() {
        if (!getPassphrase().isEmpty()) {
            return true;
        }
        return false;
    }
}
